samcli/commands/init/__init__.py

```diff
@@ -39,107 +55,57 @@
-        \b
-        $ sam init --location https://example.com/path/to/template.zip
-
-        \b
-        Initializes a new SAM project using custom template in a local path
-        \b
-        $ sam init --location /path/to/template/folder
-
-    """
-    # All logic must be implemented in the `do_cli` method. This helps ease unit tests
-    do_cli(ctx, location, runtime, dependency_manager, output_dir, name, no_input)  # pragma: no cover
-
-
-def do_cli(ctx, location, runtime, dependency_manager, output_dir, name, no_input):
-    """
-    Implementation of the ``cli`` method, just separated out for unit testing purposes
-    """
-    from samcli.commands.exceptions import UserException
-    from samcli.local.init import generate_project
-    from samcli.local.init.exceptions import GenerateProjectFailedError
-
-    LOG.debug("Init command")
-    click.secho("[+] Initializing project structure...", fg="green")
-
-    no_build_msg = """
-Project generated: {output_dir}/{name}
-
-Steps you can take next within the project folder
-===================================================
-[*] Invoke Function: sam local invoke HelloWorldFunction --event event.json
-[*] Start API Gateway locally: sam local start-api
-""".format(
-        output_dir=output_dir, name=name
-    )
-
-    build_msg = """
-Project generated: {output_dir}/{name}
-
-Steps you can take next within the project folder
-===================================================
-[*] Install dependencies
-[*] Invoke Function: sam local invoke HelloWorldFunction --event event.json
-[*] Start API Gateway locally: sam local start-api
-""".format(
-        output_dir=output_dir, name=name
-    )
-
-    no_build_step_required = (
-        "python",
-        "python3.7",
-        "python3.6",
-        "python2.7",
-        "nodejs",
-        "nodejs4.3",
-        "nodejs6.10",
-        "nodejs8.10",
-        "nodejs10.x",
-        "ruby2.5",
-    )
-    next_step_msg = no_build_msg if runtime in no_build_step_required else build_msg
-
-    try:
-        generate_project(location, runtime, dependency_manager, output_dir, name, no_input)
-        if not location:
-            click.secho(next_step_msg, bold=True)
-            click.secho("Read {name}/README.md for further instructions\n".format(name=name), bold=True)
-        click.secho("[*] Project initialization is now complete", fg="green")
-    except GenerateProjectFailedError as e:
-        raise UserException(str(e))
+def cli(ctx, no_interactive, location, runtime, dependency_manager, output_dir, name, app_template, no_input):
+    do_cli(
+        ctx, no_interactive, location, runtime, dependency_manager, output_dir, name, app_template, no_input
+    )  # pragma: no cover
+
+
+def do_cli(
+    ctx,
+    no_interactive,
+    location,
+    runtime,
+    dependency_manager,
+    output_dir,
+    name,
+    app_template,
+    no_input,
+    auto_clone=True,
+):
+    # check for mutually exclusive parameters
+    if location and app_template:
```

Comment:
can potentially lift exception cases into decorators on `do_cli`

Reply:
That turned out to be a bit problematic when I tried it, as we don't always move to failure - rather, in some cases where we have missing required parameters we move to interactive, in other cases we fail. It's difficult to lift a partial of this out without making the code look more confusing to me.